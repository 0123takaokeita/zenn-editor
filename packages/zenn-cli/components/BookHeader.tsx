import { useRef, useEffect, useState } from "react";
import { useRouter } from "next/router";
import { Book, ErrorMessage } from "@types";
import { getBookErrors } from "@utils/validator";
import { ContentWrapper } from "@components/ContentWrapper";
import { ErrorRow } from "@components/ErrorRow";

type Props = { book: Book };

export const BookHeader: React.FC<Props> = ({ book }) => {
  const [isCoverRatioError, setIsCoverRatioError] = useState<boolean>(false);
  const isCover = !!book.coverDataUrl;
  const baseErrorMessages = getBookErrors(book);
  const coverRatioError: ErrorMessage = {
    isCritical: true,
    message:
      "カバー画像の「幅 : 高さ」の比率は「1 : 1.4」にしてください（最終的に幅500px・高さ700pxにリサイズされます）",
  };

  const errorMessages = isCoverRatioError
    ? [coverRatioError, ...baseErrorMessages]
    : baseErrorMessages;
  const errorCount = errorMessages?.length;

  const { asPath } = useRouter();
  const imageRef = useRef(null);
  useEffect(() => {
    if (isCover) validateImageSize(imageRef.current);
  }, [asPath]);

  const validateImageSize = (imageEl: HTMLImageElement): void => {
    const idealApectRatio = 1.4;
    const width = imageEl.naturalWidth;
    const height = imageEl.naturalHeight;
    const aspectRatio = height / width;
    setIsCoverRatioError(Math.abs(aspectRatio - idealApectRatio) > 0.1);
  };

  return (
    <header className="content-header">
      <ContentWrapper>
        {isCover && (
          <div className="content-header__cover">
            <img src={book.coverDataUrl} ref={imageRef} />
          </div>
        )}

        <h1 className="content-header__title">{book.title || "No Title"}</h1>

        <div className="content-header__row">
          <span className="content-header__row-title">slug</span>
          <span className="content-header__row-result"> {book.slug}</span>
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">summary</span>
          <span
            className="content-header__row-result"
            style={{ whiteSpace: "pre-line" }}
          >
            {book.summary || "指定が必要です"}
          </span>
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">topics</span>
          <span className="content-header__row-result">
            {Array.isArray(book.topics) && book.topics.length
              ? book.topics.map((t, i) => (
                  <span className="content-header__topic" key={`bt${i}`}>
                    {t}
                  </span>
                ))
              : "指定が必要です"}
          </span>
        </div>

        {book.published !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">published</span>
            <span className="content-header__row-result">
              {book.published ? "✅（公開）" : "false（下書き）"}
            </span>
          </div>
        )}

        {book.price !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">price</span>
            <span className="content-header__row-result">
              {book.price?.toString()}
              {book.price === 0 ? "（無料公開）" : ""}
            </span>
          </div>
        )}

        {!!errorCount && (
          <div>
            <div className="content-header__error-title">
              {errorCount}件の修正が必要です
            </div>
            {errorMessages.map((errorMessage, index) => (
              <ErrorRow errorMessage={errorMessage} key={`invldmsg${index}`} />
            ))}
          </div>
        )}

        <a
          href="https://zenn.dev/zenn/articles/edit-book-using-cli"
          className="content-header__link"
          rel="noopener noreferrer"
          target="_blank"
        >
          Bookの作成方法はこちら →
        </a>
      </ContentWrapper>
    </header>
  );
};
