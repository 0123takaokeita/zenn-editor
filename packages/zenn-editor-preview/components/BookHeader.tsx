import { Book } from "@types";
import { getBookErrors } from "@utils/validator";
import ContentWrapper from "@components/ContentWrapper";
import ErrorRow from "@components/ErrorRow";

const BookHeader: React.FC<{ book: Book }> = ({ book }) => {
  const errorMessages = getBookErrors(book);
  const errorCount = errorMessages?.length;
  return (
    <header className="content-header">
      <ContentWrapper>
        <h1 className="content-header__title">{book.title || "No Title"}</h1>

        <div className="content-header__row">
          <span className="content-header__row-title">slug</span>
          {book.slug}
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">summary</span>
          {book.summary || "指定が必要です"}
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">topics</span>
          {Array.isArray(book.topics)
            ? book.topics.map((t, i) => (
                <span className="content-header__topic" key={`bt${i}`}>
                  {t}
                </span>
              ))
            : "指定が必要です"}
        </div>

        {book.public !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">public</span>
            {book.public?.toString()}
            {book.public === false ? "（非公開）" : ""}
          </div>
        )}

        {book.price !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">price</span>
            {book.price?.toString()}
            {book.price === 0 ? "（無料公開）" : ""}
          </div>
        )}

        {!!errorCount && (
          <div>
            <div className="content-header__error-title">
              {errorCount}件のエラー
            </div>
            {errorMessages.map((errorMessage, index) => (
              <ErrorRow errorMessage={errorMessage} key={`invldmsg${index}`} />
            ))}
          </div>
        )}

        <a href="todo" className="content-header__link" target="_blank">
          Bookの作成方法はこちら →
        </a>
      </ContentWrapper>
    </header>
  );
};
export default BookHeader;
