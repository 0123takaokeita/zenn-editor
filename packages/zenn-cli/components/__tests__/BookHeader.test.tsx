import React from "react";
import { Book } from "@types";
import { createRenderer } from "react-test-renderer/shallow";
import { BookHeader } from "@components/BookHeader";

const renderer = createRenderer();
const useRouter = jest.spyOn(require("next/router"), "useRouter");

const dummyBook: Book = {
  chapters: [],
  coverDataUrl: "",
  price: 100,
  published: false,
  slug: "dummy-book-test",
  summary: "OK\nOK\nOK",
  title: "test book",
  topics: ["zenn"],
};
describe("<BookHeader/>", () => {
  it("should render and match the snapshot", () => {
    // mock useRouter()
    // See this link: https://github.com/vercel/next.js/issues/7479
    useRouter.mockImplementationOnce(() => ({
      query: {},
    }));
    renderer.render(<BookHeader book={dummyBook} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
