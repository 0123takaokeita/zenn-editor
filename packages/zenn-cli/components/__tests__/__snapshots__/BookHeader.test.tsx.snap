// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookHeader/> should render and match the snapshot 1`] = `
<header
  className="content-header"
>
  <Unknown>
    <h1
      className="content-header__title"
    >
      test book
    </h1>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        slug
      </span>
      <span
        className="content-header__row-result"
      >
         
        dummy-book-test
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        summary
      </span>
      <span
        className="content-header__row-result"
        style={
          Object {
            "whiteSpace": "pre-line",
          }
        }
      >
        OK
OK
OK
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        topics
      </span>
      <span
        className="content-header__row-result"
      >
        <span
          className="content-header__topic"
        >
          zenn
        </span>
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        published
      </span>
      <span
        className="content-header__row-result"
      >
        false（下書き）
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        price
      </span>
      <span
        className="content-header__row-result"
      >
        100
        
      </span>
    </div>
    <div>
      <div
        className="content-header__error-title"
      >
        2
        件の修正が必要です
      </div>
      <Unknown
        errorMessage={
          Object {
            "isCritical": true,
            "message": "price（本の価格）を有料にする場合、200〜5000の間で指定してください",
          }
        }
      />
      <Unknown
        errorMessage={
          Object {
            "isCritical": false,
            "message": "本のカバー画像（cover.pngもしくはcover.jpg）を「/books/dummy-book-test」ディレクトリ内に配置してください",
          }
        }
      />
    </div>
    <a
      className="content-header__link"
      href="https://zenn.dev/zenn/articles/zenn-cli-guide#cliで本（book）を管理する"
      rel="noopener noreferrer"
      target="_blank"
    >
      本の作成方法 →
    </a>
  </Unknown>
</header>
`;
